### ISSUE TITLE
Racy controller from framework.NewInformer?

### ISSUE THREAD

**Issue by jsafrane** [kubernetes/kubernetes#26082]
When I stress-test PersistentVolume controller I sometimes get (correct) "PersistentVolumeClaim deleted" event followed by (wrong) "PersistentVolumeClaim added". This happens very rarely (<0.01%) when I delete a PersistentVolumeClaim and the controller is under heavy load.

I use `framework.NewInformer` ([link](https://github.com/jsafrane/kubernetes/blob/master/pkg/controller/framework/controller.go#L213) ) like everybody else. Reading its code, what happens when the controller is processing 'PVC deleted' event at line [251](https://github.com/jsafrane/kubernetes/blob/master/pkg/controller/framework/controller.go#L251), i.e just **before** `Delete()` is called and at the same time periodic sync starts? The PVC that is being deleted is still in the cache and a 'PVC Sync' will be added into `Queue` and my controller gets  `OnDelete()` callback followed by `OnAdd()`.

My PersistentVolume controller then sees an old claim and tries to do stuff with it and sometimes it updates the claim in etcd. So the claim gets 'resurrected' in etcd even though user deleted it.

It is possible that my controller does something terribly wrong - I checked it twice, but it's pretty complicated piece of code. I am 100% sure that `OnAdd()` is called as reaction to `cache.Sync` event, I did a debug log in `Process` function of Informer.

**Comment by jsafrane** [comment]
@lavalamp, you're author of the potentially racy code, can you please take a look?

**Comment by jsafrane** [comment]
I have reproduced of the observed behavior. Get it at [reproduce-informer-race branch](https://github.com/jsafrane/kubernetes/tree/reproduce-informer-race) of my Kubernetes fork, patch https://github.com/jsafrane/kubernetes/commit/0b5bc1dc2fb1cc7c1e5fdb0d7cf4878f298d1b85

Compile it:
```
cd test/integration/
godep go test -tags 'integration no-docker' -c
```

Run etcd somewhere and then run the test:
```
./integration.test -test.run TestInformerRace -v 2
```
(with `-v 3` you get some logs I added to the informer and reflector)

The test spawns a very simple controller with `AddFunc`, `ModifyFunc` and `DeleteFunc` and creates, modifies and deletes one PV. `ModifyFunc` is intentionally slow to trigger periodic sync when the PV was already deleted by the client, but this delete event was not yet processed by the controller - the volume is still in the controller cache. This means in the controller queue is `cache.Deleted` and `cache.Sync` **after** that. `cache.Deleted` event removes the volume from cache and `cache.Sync` does not find it there, therefore invokes `AddFunc` callback and adds the PV to the cache -> troubles.

**Comment by jsafrane** [comment]
I have sort of fix, please review very carefully #26167

**Comment by lavalamp** [comment]
I was pretty sure there was a race in there, but never able to repro, thanks.

**Comment by jsafrane** [comment]
Fixed by #27341
